File: leetcode/349-intersection-of-two-arrays/IntersectionOfTwoArrays.java
class Solution {
    public int[] intersection(int[] nums1, int[] nums2) {
        Set<Integer> used = new HashSet<>();
        
        for (int num : nums1) {
            used.add(num);
        }
        
        Set<Integer> resultSet = new HashSet<>();
        for (int num : nums2) {
            if (used.contains(num)) {
                resultSet.add(num);
            }
        }
        
        int[] result = new int[resultSet.size()];
        
        int i = 0;
        for (int num : resultSet) {
            result[i++] = num;
        }
        
        return result;
    }
}

File: leetcode/349-intersection-of-two-arrays/intersection-of-two-arrays.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
const intersection = function (nums1, nums2) {
  const set = new Set();

  for (const num of nums1) {
    set.add(num);
  }

  const result = new Set();
  for (const num of nums2) {
    if (set.has(num)) {
      result.add(num);
    }
  }

  return [...result];
};
